src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -166,6 +179,221 @@ instance (NFData t, NFData r) => NFData (SnapMergingRunState t r) where
+              -- have to redo merging work here. SuppliedCredits tracks how
+              -- many credits were supplied before the snapshot was taken.
+              leftoverCredits <- MR.supplyCredits
+                                   mr (creditThresholdForLevel conf ln)
+                                   (MR.Credits sc)
+              assert (leftoverCredits == 0) $ return mr
+
+    fromSnapPreExistingRun :: SnapPreExistingRun (Ref (Run m h)) -> m (MT.PreExistingRun m h)
+    fromSnapPreExistingRun (SnapPreExistingRun run) = MT.PreExistingRun <$> dupRun run
+    fromSnapPreExistingRun (SnapPreExistingMergingRun nr ne (SuppliedCredits sc) smrs) =
+        MT.PreExistingMergingRun <$> getSnapMergingRunState nr ne sc smrs
+
+
+{-------------------------------------------------------------------------------
+  Conversion to merge tree snapshot format
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE toSnapMergingTree :: MT.MergingTree IO h -> IO (SnapMergingTree (Ref (Run IO h))) #-}
+toSnapMergingTree ::
+     (PrimMonad m, MonadMVar m)
+  => MT.MergingTree m h
+  -> m (SnapMergingTree (Ref (Run m h)))
+toSnapMergingTree (MT.MergingTree mStateVar _mCounter) =
+  withMVar mStateVar $ \mState -> SnapMergingTree <$> toSnapMergingTreeState mState
+
+{-# SPECIALISE toSnapMergingTreeState :: MT.MergingTreeState IO h -> IO (SnapMergingTreeState (Ref (Run IO h))) #-}
+toSnapMergingTreeState ::
+     (PrimMonad m, MonadMVar m)
+  => MT.MergingTreeState m h
+  -> m (SnapMergingTreeState (Ref (Run m h)))
+toSnapMergingTreeState (MT.CompletedTreeMerge r) = pure $ SnapCompletedTreeMerge r
+toSnapMergingTreeState (MT.PendingTreeMerge p) = SnapPendingTreeMerge <$> toSnapPendingMerge p
+toSnapMergingTreeState (MT.OngoingTreeMerge mergingRun) = do
+    -- We need to know how many credits were supplied so we can restore merge
+    -- work on snapshot load.
+    (mergingRunState,
+     MR.SuppliedCredits (MR.Credits suppliedCredits),
+     mergeNumRuns,
+     mergeNumEntries) <- MR.snapshot mergingRun
+    -- TODO: MR.snapshot needs to return duplicated run references, and we
+    -- need to arrange to release them when the snapshoting is done.
+    let smts = toSnapMergingRunState mergingRunState
+    pure $
+        SnapOngoingTreeMerge
+        mergeNumRuns
+        mergeNumEntries
+        (SuppliedCredits suppliedCredits)
+        smts
+
+{-# SPECIALISE toSnapPendingMerge :: MT.PendingMerge IO h -> IO (SnapPendingMerge (Ref (Run IO h))) #-}
+toSnapPendingMerge ::
+     (PrimMonad m, MonadMVar m)
+  => MT.PendingMerge m h
+  -> m (SnapPendingMerge (Ref (Run m h)))
+toSnapPendingMerge (MT.PendingUnionMerge mts) =
+  SnapPendingUnionMerge <$> traverse (\(DeRef mTree) -> toSnapMergingTree mTree) mts
+toSnapPendingMerge (MT.PendingLevelMerge pes mmt) = do
+  pes' <- traverse toSnapPreExistingRun pes
+  mmt' <- traverse (\(DeRef mTree) -> toSnapMergingTree mTree) mmt
```

Comment:
Looks like it might be better for `toSnapMergingTree` to take a `Ref` and do the `DeRef`` itself, rather than here in the callers.